### PR TITLE
support obtaining a build env with nix

### PR DESCRIPTION
**Pull request by kazcw**
### Pull Request Description

This packages a build environment for the Nix package manager. I implemented this over the last couple of days in order to get Enso building on my computer. I'm nominating it for inclusion in the repo since it allows any future developer who has Nix to get started working on Enso with a single standard command.

### Important Notes

There are two different .nix scripts:
- shell.nix: provides an environment suitable for building the IDE and Engine
- shell-fhs.nix: provides an environment suitable for building the IDE, and using the binary Engine that is downloaded by default

### Checklist

- [x] The documentation has been updated if necessary. (It is self-documenting; shell.nix has a standard meaning to nix users)
- [x] All code conforms to the [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md), [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md), and [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guides.
- [x] All documentation and configuration conforms to the [markdown](https://github.com/enso-org/enso/blob/develop/docs/style-guide/markdown.md) and [YAML](https://github.com/enso-org/enso/blob/develop/docs/style-guide/yaml.md) style guides.
- [x] All code has been tested where possible.

### CODE REVIEW

File: nix/wasm-pack_0_9/default.nix [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+, rustPlatform
+, pkg-config
+, libressl
+, curl
+, Security
+}:
+
+rustPlatform.buildRustPackage rec {
+  pname = "wasm-pack";
+  version = "0.9.1";
```
Comment (4e6) [inline nix/wasm-pack_0_9/default.nix]: I believe we can use the latest wasm-pack 0.10.2 https://github.com/enso-org/enso/blob/a74e075adca3518cf57d0d78380b083e93f4dec7/app/gui/docs/CONTRIBUTING.md#development-environment

Comment (vitvakatu) [reply nix/wasm-pack_0_9/default.nix]: Yeah, we actually must use the latest one to avoid issues on M1 Macs

**Comment by mwu-tow** [comment]
To be honest, at this point I'm not convinced that this should be a part of this repository.

This adds yet another place where we hardcode versions of our dev dependencies. I like to keep them to the reasonable minimum.
The code is non-trivial and limited to a specific development environment that no-one else in the team uses (I think?) — and as such is prone to get broken when other developers change things. The very minimum to help this would be setting up a CI workflow that checks if this setup can successfully build the packages. But that would likely run against the current efforts to make CI run faster. And for example I don't think I have domain knowledge to be comfortable with editing such configs.

I have also some concerns about the approach "copy the official package and patch" approach. This looks like another place where things could diverge. Could these patches be somehow upstreamed?

My current CI work should eventually provide this repo with Docker images for Windows and Linux-based environments. While this is somewhat a blunt weapon, it should cover the case of developer wanting to build things with minimal manual environment configuration effort. As at the same time the environment will be used by CI, its being up to date will be enforced with no additional overhead.

**Review by mwu-tow — CHANGES_REQUESTED** [review]
Comment above ^.

**Comment by akavel** [comment]
For what it's worth, I'm a +1 person interested in this (as an active user of Nix, also at work); with that said, I absolutely understand various tradeoffs connected to having to maintain an extra script in a not-yet-widely known technology, etc.

With that said, I'm kind of wondering if there could be a way to include it somehow in a half-official way? For example, the Go language repository used to include a `contrib/` directory, which contained for example user-contributed Vim plugins, and now includes a `misc/` directory with some more or less garden-variety stuff (https://github.com/golang/go/tree/master/misc).

<blockquote><div> GitHub</div><div><strong><a href="https://github.com/golang/go">go/misc at master · golang/go</a></strong></div><div>The Go programming language. Contribute to golang/go development by creating an account on GitHub.</div></blockquote>

**Comment by kazcw** [comment]
<details><summary>(hidden: replies to feedback details)</summary>

> This adds yet another place where we hardcode versions of our dev dependencies. I like to keep them to the reasonable minimum.
> The code is non-trivial and limited to a specific development environment that no-one else in the team uses (I think?) — and as such is prone to get broken when other developers change things. The very minimum to help this would be setting up a CI workflow that checks if this setup can successfully build the packages. But that would likely run against the current efforts to make CI run faster. And for example I don't think I have domain knowledge to be comfortable with editing such configs.

As far as the maintenance burden, maintaining this is how I keep my development environment up to date. On NixOS, rather than updating some systemwide packages when dependencies change, I update this .nix file, and that determines what dependencies are present in my Enso build environment. However I agree that it would be problematic if the nix script would not be up to date with dependencies in the time between when a change to dependencies is committed and a nix user updates the script to match.

> I have also some concerns about the approach "copy the official package and patch" approach. This looks like another place where things could diverge. Could these patches be somehow upstreamed?

Nixpkgs does support coexisting versions of some packages, but only Enso needs this particular version of GraalVM. The Nix Graal package maintainers probably wouldn't accept a PR for Enso's Graal version unless I volunteered to maintain it, and then I don't think we'd be gaining anything. However there is not much maintenance to be done: changes should only be necessary when Enso's dependency requirements change.

</details>

> To be honest, at this point I'm not convinced that this should be a part of this repository.

After considering your feedback, I agree that the nix script should not be in-repo unless it is kept up to date and checked with CI, and doing that would have excess overhead if it adds a new set of hardcoded version numbers that require Nix knowledge to update.

So, I think a better approach would be:
- modify the nix script to be usable as a Nixpkgs overlay
- host the overlay in a separate repository
- link to the overlay in the developer docs somewhere

I'll update the plan in the Pivotal story.

**Pull request closed without merging**